geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/index/BucketIndex.scala

```diff
@@ -20,20 +22,22 @@ class BucketIndex[T](xBuckets: Int = 360,
                      yBuckets: Int = 180,
                      extents: Envelope = new Envelope(-180.0, 180.0, -90.0, 90.0)) extends SpatialIndex[T] {
 
-  // create the buckets upfront to avoid having to synchronize the whole array
   private val gridSnap = new GridSnap(extents, xBuckets, yBuckets)
-  private val buckets = Array.fill(xBuckets, yBuckets)(mutable.HashSet.empty[T])
+  // create the buckets up front to avoid having to synchronize the whole array
+  // we use a concurrentHashSet, which gives us iterators that aren't affected by modifications to the backing set
+  private def newSet = Collections.newSetFromMap(new ConcurrentHashMap[T, java.lang.Boolean])
+  private val buckets = Array.fill(xBuckets, yBuckets)(newSet)
 
   override def insert(envelope: Envelope, item: T): Unit = {
     val (i, j) = getBucket(envelope)
     val bucket = buckets(i)(j)
-    bucket.synchronized(bucket.add(item))
+    bucket.add(item)
```

Comment:
i'm guessing the main performance boost will be from line 65. Thanks for the reminder about ConcurrentHashMaps, btw :)